### PR TITLE
Migrate write ahead log

### PR DESCRIPTION
**Pull request by fabxc**
On startup, rewrite the old write ahead log into the new format once.

@jkohen @brian-brazil

### CODE REVIEW

File: wal.go [hunk 1]
```diff
@@ -1206,3 +1211,89 @@ func (r *walReader) decodeDeletes(flag byte, b []byte, res *[]Stone) error {
...
+			err = repl.Log(enc.Tombstones(s, b[:0]))
+		},
+	)
+	if decErr != nil {
+		return errors.Wrap(err, "decode old entries")
+	}
+	if err != nil {
+		return errors.Wrap(err, "write new entries")
+	}
+	if err := w.Close(); err != nil {
```
Comment (jkohen) [inline wal.go]: Don't you want to close both WALs regardless of errors?

**Comment by krasi-georgiev** [comment]
I was thinking that with more versions and more changes this repair/convert blocks might "contaminate" :hankey:  the code. Why not add this as part of the tsdb cli tool?

I have already  added [tsdb scan](https://github.com/prometheus/tsdb/pull/320) so  maybe something like `tsdb upgrade`

**Comment by fabxc** [comment]
Because our stability guarantee is that you can just restart Prometheus
with the next minor and things will work. Telling people to use a CLI tool
is breaking that guarantee.

**Comment by fabxc** [comment]
Rebased and resolved conflicts.

**Comment by fabxc** [comment]
Rebased and resolved conflicts.

File: wal.go [hunk 2]
```diff
@@ -173,6 +176,8 @@ func newCRC32() hash.Hash32 {
 }
 
 // SegmentWAL is a write ahead log for series data.
+//
+// DEPRECATED: use wal pkg combined with the record coders instead.
```
Comment (brian-brazil) [inline wal.go]: s/coders/codex/

File: wal.go [hunk 3]
```diff
@@ -1206,3 +1211,100 @@ func (r *walReader) decodeDeletes(flag byte, b []byte, res *[]Stone) error {
...
+
+	tmpdir := dir + ".tmp"
+	if err := os.RemoveAll(tmpdir); err != nil {
+		return errors.Wrap(err, "cleanup replacement dir")
+	}
+	repl, err := wal.New(logger, nil, tmpdir)
+	if err != nil {
+		return errors.Wrap(err, "open new WAL")
+	}
+	// We close it once already before as part of finalization.
```
Comment (brian-brazil) [inline wal.go]: This doesn't read properly

File: wal.go [hunk 4]
```diff
@@ -1206,3 +1211,100 @@ func (r *walReader) decodeDeletes(flag byte, b []byte, res *[]Stone) error {
...
+	if decErr != nil {
+		return errors.Wrap(err, "decode old entries")
+	}
+	if err != nil {
+		return errors.Wrap(err, "write new entries")
+	}
+	if err := repl.Close(); err != nil {
+		return errors.Wrap(err, "close new WAL")
+	}
+	if err := fileutil.Rename(tmpdir, dir); err != nil {
```
Comment (brian-brazil) [inline wal.go]: This no longer works given that dir already exists and Rename no longer deletes the destination first.

**Comment by fabxc** [comment]
Merging this into the branch for the new WAL so we can have a final pass over all changes together.

**Pull request merged**